game-core/src/test/java/games/strategy/triplea/attachments/TriggerAttachmentTest.java

```diff
@@ -26,102 +29,226 @@
+
+      final ProductionFrontierList productionFrontierList = gameData.getProductionFrontierList();
+      productionFrontierList
+          .addProductionFrontier(
+              new ProductionFrontier("frontier", gameData, Collections.singletonList(productionRule2)));
+
+      TriggerAttachment.triggerProductionFrontierEditChange(
+          satisfiedTriggers,
+          bridge,
+          "beforeOrAfter",
+          "stepName",
+          false, // useUses
+          false, // testUses
+          false, // testChance
+          false); // testWhen
+      final ArgumentCaptor<CompositeChange> argument = ArgumentCaptor.forClass(CompositeChange.class);
+      verify(bridge).addChange(argument.capture());
+      final CompositeChange change = argument.getValue();
+      assertFalse(change.isEmpty());
+      final ArgumentCaptor<String> ruleAddArgument = ArgumentCaptor.forClass(String.class);
+      verify(historyWriter, times(3)).startEvent(ruleAddArgument.capture());
+      final List<String> allValues = ruleAddArgument.getAllValues();
+      assertEquals(3, allValues.size());
+      assertTrue(allValues.stream().anyMatch(s -> s.contains("rule1") && s.contains("added")));
+      assertTrue(allValues.stream().anyMatch(s -> s.contains("rule2") && s.contains("removed")));
+      assertTrue(allValues.stream().anyMatch(s -> s.contains("rule3") && s.contains("added")));
+      assertTrue(allValues.stream().allMatch(s -> s.contains("frontier")));
+    }
+
+    @Test
+    void testTriggerPlayerPropertyChange() throws Exception {
+      final GameData gameData = bridge.getData();
+      final TriggerAttachment triggerAttachment = mock(TriggerAttachment.class);
+      final Set<TriggerAttachment> satisfiedTriggers = Collections.singleton(triggerAttachment);
+
+      when(triggerAttachment.getPropertyMap()).thenCallRealMethod();
+      triggerAttachment.getPropertyMap().get("playerAttachmentName").setValue("rulesAttachment:RulesAttachment");
+      when(triggerAttachment.getPlayerProperty())
+          .thenReturn(Collections.singletonList(
+              Tuple.of("productionPerXTerritories", "someNewValue")));
+      when(triggerAttachment.getName()).thenReturn("mockedTriggerAttachment");
+
+      final PlayerId playerId = mock(PlayerId.class);
+      when(playerId.getAttachment("rulesAttachment"))
+          .thenReturn(new RulesAttachment(null, null, gameData));
+      when(triggerAttachment.getPlayers()).thenReturn(Collections.singletonList(playerId));
+
+      TriggerAttachment.triggerPlayerPropertyChange(
+          satisfiedTriggers,
+          bridge,
+          "beforeOrAfter",
+          "stepName",
+          false, // useUses
+          false, // testUses
+          false, // testChance
+          false); // testWhen
+      verify(bridge).addChange(not(argThat(Change::isEmpty)));
+    }
+
+    // TODO: Any existing games using 'triggerRelationshipTypePropertyChange', to base the test parameters on?
```

Comment:
Doesn't look like it. https://github.com/search?q=org%3Atriplea-maps+triggerRelationshipTypePropertyChange&unscoped_q=triggerRelationshipTypePropertyChange

I wonder if we should keep that trigger @ron-murhammer

Reply:
You'll want to search for 'relationshipTypeProperty' in map XML but it only appears in POS2. Its probably in some of the larger ones which github doesn't index as they are over the max default github file size for searching.

Reply:
Thanks, I think I understand that part of the code better, and I refactored it and added a unit test for it.